src/library/scala/collection/Seq.scala

```diff
@@ -691,20 +691,16 @@ trait SeqOps[+A, +CC[_], +C] extends Any
     *              sorted according to the ordering `ord`.
     */
   def sorted[B >: A](implicit ord: Ordering[B]): C = {
-    val len = this.length
+    val known = this.knownSize
     val b = newSpecificBuilder
-    if (len == 1) b ++= toIterable
-    else if (len > 1) {
+    if (known == 1) b += head
+    else if (known != 0) {
+      val arr = toArray[AnyRef]
```

Comment:
this does not work because there's no guarantee that `AnyRef >: A`. is there a way around that without copying and casting each element individually?

alternatively, we could create an `ArrayBuffer` from the elements regardless, and then sort it in place (which incidentally, should use `java.util.Arrays.sort` rather than `scala.util.Sorting.stableSort`, since `ArrayBuffer` is always backed by `Array[AnyRef]`, and `Sorting.stableSort` uses `Arrays.sort` in that case). and then we could straight up use the `ArrayBuffer` as the underlying `Seq`

Reply:
The default erased representation of `Array[Any]` is `Array[Object]` so it should be safe to do `toArray[Any]`. Collections that are based on primitive arrays should override methods like `sorted` so it's a fair assumption that data is already boxed, therefore we should keep it boxed rather than unboxing primitive types. And we don't have a way of getting an array with the native representation from an arbitrary collection so there's not much we can do about it anyway.

I think it's even safe to cast the result of `toArray[Any]` back to `Array[AnyRef]` for more efficient access. Scala arrays are invariant after all. While an `Array[T] forSome { type T }` that erases to `Array[Any]` could be a primitive array, an actual `Array[Any]` that you requested by passing in a `ClassTag[Any]` should never be primitive.

I'd also keep the `length` call here in `SeqOps`. It's probably better (but at least much simpler and not much worse) to traverse a linear Seq twice than to perform an additional array copy.

Reply:
what's wrong with the `length` call in `SeqView` then? they both act on the underlying `Seq`

Reply:
I had to take another look at `SeqView` because I remembered it wrong. Any expensive operation that requires a full traversal to compute the length only results in a `View`, so it should indeed be okay to call `length` in `SeqView`.

Reply:
@szeiger the only thing I'm concerned about is if other platforms (e.g. scala-js) will have problems with casting `Array[Any]` to `Array[AnyRef]`

Reply:
is it also safe to do it the other way? i.e. could we use any `Array[AnyRef]`, but cast it to `Array[Any]` just to copy elements into it?

Reply:
Actually, using `Array[Any]` everywhere should be sufficient. An explicit `Array[Any]` always erases to `Array[Object]` and the compiler doesn't generate code for polymorphic arrays:

```
def f(xs: Array[Any]): Any = xs(1)

         0: aload_1
         1: iconst_1
         2: aaload
         3: areturn
```

You only get that for polymorphic arrays with a lower bound of `Any`:

```
def f(xs: Array[_]): Any = xs(1)

         0: getstatic     #22                 // Field scala/runtime/ScalaRunTime$.MODULE$:Lscala/runtime/ScalaRunTime$;
         3: aload_1
         4: iconst_1
         5: invokevirtual #26                 // Method scala/runtime/ScalaRunTime$.array_apply:(Ljava/lang/Object;I)Ljava/lang/Object;
         8: areturn
```